In BC event-loop, only run tasks related to fully-active documents

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #22792 (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22802)
<!-- Reviewable:end -->

gterzian committed:
in BC event-loop, only run tasks related to fully-active documents

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/script_thread.rs, components/script/task_queue.rs, components/script/dom/serviceworkerglobalscope.rs, components/script/dom/dedicatedworkerglobalscope.rs
 * @kichjang: components/script/script_thread.rs, components/script/task_queue.rs, components/script/dom/serviceworkerglobalscope.rs, components/script/dom/dedicatedworkerglobalscope.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -881,6 +894,21 @@ impl ScriptThread {
+                        inactive.insert(id.clone());
+                    }
+                }
+                inactive
```
```rust
script_thread
    .documents
    .borrow()
    .iter()
    .filter_map(|(id, document)| {
        if !document.is_fully_active() {
            Some(id.clone())
        } else {
            None
        }
    })
    .collect()
```

gterzian replied:
thanks

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
This one is ready for review. 

The test I added is not passing because of https://github.com/servo/servo/issues/22815 (it was passing locally after some manual hacks to prevent the pipeline from being closed).

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
This one is ready for review. 

The test I added is not passing because of https://github.com/servo/servo/issues/22815 (it was passing locally after some manual hacks to prevent the "active" pipeline from being closed).

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
This one is ready for review. 

The test I added is not passing because of https://github.com/servo/servo/issues/22815 (it was passing locally after some manual hacks to prevent the "active" pipeline from being closed).

gterzian commented:
Should we try to move forward on this one? I'm in no hurry otherwise 😄 

 @nox you were randomly assigned, perhaps you're interested? @jdm you reviewed the initial task queue code so this might be a good fit for you?

Note that the test fails because of https://github.com/servo/servo/issues/22815, although I could get it to pass by basically removing the code that closes pipelines. 

So in a way this PR isn't going to have much effect until we address the other issue, and it might still be a good idea to merge this in the mean-time, since it's not clear yet how the other issue will be fixed.

jdm commented:
@nox review ping!

nox commented:
Is this really not currently tested at all upstream?

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9b65bf9cc9db81b219374d35f956d6c2463b2ca7 with merge dd6e5e27dedbf265e24fb144e2519c5ab0369908...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11683)

gterzian commented:
```
  ▶ CRASH [expected OK] /html/browsers/history/the-history-interface/history_pushstate_nooptionalparam.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ CloneApi message response was dropped while Webrender was still alive: io error: All senders for this socket closed (thread main, at /home/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/2ff8da0/webrender_api/src/api.rs:936)
```

seems unrelated

gterzian commented:
@nox I couldn't find a test upstream, although it might be tested indirectly in a few other tests, I've also just asked at https://github.com/whatwg/html/issues/4242

nox commented:
Is the unexpected `PASS` of `/workers/semantics/navigation/002.html` unrelated?

nox commented on components/script/dom/dedicatedworkerglobalscope.rs:
```diff
@@ -131,6 +137,11 @@ impl QueuedTaskConversion for DedicatedWorkerScriptMsg {
 
+    fn inactive_msg() -> Self {
+        // Inactive is only relevant in the context of a browsing-context event-loop.
+        unreachable!("Workers should never receive messages marked as inactive");
```
Nit: make that `panic`, not `unreachable`, the latter is for self-contained code for which you can locally check that the invariant holds.

nox commented on components/script/dom/serviceworkerglobalscope.rs:
```diff
@@ -85,6 +92,11 @@ impl QueuedTaskConversion for ServiceWorkerScriptMsg {
 
+    fn inactive_msg() -> Self {
+        // Inactive is only relevant in the context of a browsing-context event-loop.
+        unreachable!("Workers should never receive messages marked as inactive");
```
`panic`

nox commented on components/script/script_thread.rs:
```diff
@@ -289,6 +291,19 @@ impl QueuedTaskConversion for MainThreadScriptMsg {
+            CommonScriptMsg::Task(_category, _boxed, pipeline_id, _task_source) => {
+                pipeline_id.as_ref()
+            },
+            _ => return None,
```
Nit: just `None` is ok here.

nox commented on components/script/script_thread.rs:
```diff
@@ -881,6 +900,29 @@ impl ScriptThread {
         })
     }
 
+    pub fn get_currently_not_fully_active_document_ids() -> HashSet<PipelineId> {
```
Could you explain why a `HashSet` of non-fully active documents is needed here? Aren't there less full active documents than non-fully active ones at any given time?

nox commented on components/script/task_queue.rs:
```diff
@@ -53,22 +58,85 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
+            if currently_inactive.contains(pipeline_id) {
+                let inactive_queue = inactive
+                    .entry(pipeline_id.clone())
+                    .or_insert(VecDeque::new());
```
Nit: `HashMap::or_default`

nox commented on components/script/task_queue.rs:
```diff
@@ -53,22 +58,85 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
+            inactive.keys().map(|id| id.clone()).collect()
+        };
+        let mut inactive = self.inactive.borrow_mut();
+        for pipeline_id in previously_inactive.difference(&currently_inactive) {
```
AFAICT, there is no need to collect the inactive IDs, just loop over the entries of `self.inactive` and check that the key isn't in `currently_inactive`.

nox commented on components/script/task_queue.rs:
```diff
@@ -53,22 +58,85 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
+        let mut incoming = vec![];
+
+        // 1. Make any previously stored task from now fully-active document available.
+        self.release_tasks_for_fully_active_documents(currently_inactive, &mut incoming);
```
This seems to be the only use of `release_tasks_for_fully_active_documents`, is there any reason it's passed a mutable `Vec<_>` instead of returning one?

nox reviewed (commented):
Some nits and remarks but it seems good otherwise.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
> Is the unexpected `PASS` of `/workers/semantics/navigation/002.html` unrelated?

I think so. Didn't get a change in last run to see if it happened on other boxes...

Comments addressed, and I would like to see if the test still can passes, which require some manual local testing which I haven't had time to do yet. 

It seems that there is no dedicated upstream test.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 5dda9355da203f92db366f10dc608e968cfd239b with merge 8b2b2a7f91af2ee9e8a6055fff1fc740923d002c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11760)

CYBAI commented:
```
error: unused variable: `inactive_queue`
   --> components/script/task_queue.rs:103:36
    |
103 |             .filter(|(pipeline_id, inactive_queue)| fully_active.contains(pipeline_id))
    |                                    ^^^^^^^^^^^^^^ help: consider prefixing with an underscore: `_inactive_queue`
    |
    = note: `-D unused-variables` implied by `-D warnings`

error: aborting due to previous error

error: Could not compile `script`.
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 0ae51fb6f880e61218805380269c06b18c76204c with merge 6371dc989a11fe48ff25fa2d52a4b885944259b1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11710)

gterzian commented:
```
CRASH [expected TIMEOUT] /html/webappapis/dynamic-markup-insertion/opening-the-input-stream/tasks.window.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ Spurious wake-up of the event-loop, task-queue has no tasks available: () (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at src/libcore/result.rs:997)
```

Will look into...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit d9e90f187bb380853860f0fec1d85d93e570aa8b with merge c63513296fc0ebeb06e327b82c220dcfbee3846f...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#15224.

gterzian commented:
@nox Ok I fixed the previous problems, and updated the expectations to pass for `/workers/semantics/navigation/002.html`. The test I wrote, and initially expected to fail, now passes, although I am not sure why. From print statements I can tell that the logic still works as expected, did anything change recently about how we close pipelines?

nox commented:
Great work for `/workers/semantics/navigation/002.html`!

@bors-servo r+

bors-servo commented:
:pushpin: Commit ecfb9c6 has been approved by `nox`

<!-- @bors-servo r=nox ecfb9c639ad610d406e7dd278dfc7aac964ddca8 -->

bors-servo commented:
:hourglass: Testing commit ecfb9c639ad610d406e7dd278dfc7aac964ddca8 with merge f2ea73f360f30c3aad306bfc53bc8c7041088729...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/1190), [arm32](http://build.servo.org/builders/arm32/builds/14066), [arm64](http://build.servo.org/builders/arm64/builds/14094), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11734), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11790), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6672), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6640), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10882), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10946), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6696), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6627), [magicleap](http://build.servo.org/builders/magicleap/builds/809), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/LtHS_hAYTbGsQhkp3vFElw)
Approved by: nox
Pushing f2ea73f360f30c3aad306bfc53bc8c7041088729 to master...

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1552326896434.
